pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/consumer/PulsarKafkaConsumer.java

```diff
@@ -0,0 +1,461 @@
+
+        String serviceUrl = config.getList(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG).get(0);
+
+        Properties properties = new Properties();
+        config.originals().forEach((k, v) -> properties.put(k, v));
+        ClientConfiguration clientConf = PulsarKafkaConfig.getClientConfiguration(properties);
+        try {
+            client = PulsarClient.create(serviceUrl, clientConf);
+        } catch (PulsarClientException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void received(org.apache.pulsar.client.api.Consumer consumer, Message msg) {
+        // Block listener thread if the application is slowing down
+        try {
+            receivedMessages.put(new QueueItem(consumer, msg));
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public Set<TopicPartition> assignment() {
+        throw new UnsupportedOperationException("Cannot access the partitions assignements");
+    }
+
+    /**
+     * Get the current subscription. Will return the same topics used in the most recent call to
+     * {@link #subscribe(Collection, ConsumerRebalanceListener)}, or an empty set if no such call has been made.
+     *
+     * @return The set of topics currently subscribed to
+     */
+    @Override
+    public Set<String> subscription() {
+        return consumers.keySet().stream().map(TopicPartition::topic).collect(Collectors.toSet());
+    }
+
+    @Override
+    public void subscribe(Collection<String> topics) {
+        topics.forEach(topic -> {
+            try {
+                // Create individual subscription on each partition, that way we can keep using the
+                // acknowledgeCumulative()
+                PartitionedTopicMetadata partitionMetadata = ((PulsarClientImpl) client)
+                        .getPartitionedTopicMetadata(topic).get();
+
+                ConsumerConfiguration conf = new ConsumerConfiguration();
+                conf.setSubscriptionType(SubscriptionType.Failover);
+                conf.setMessageListener(this);
+                if (partitionMetadata.partitions > 1) {
+                    // Subscribe to each partition
+                    conf.setConsumerName(ConsumerName.generateRandomName());
+                    List<CompletableFuture<org.apache.pulsar.client.api.Consumer>> futures = new ArrayList<>();
+                    for (int i = 0; i < partitionMetadata.partitions; i++) {
+                        String partitionName = DestinationName.get(topic).getPartition(i).toString();
+                        futures.add(client.subscribeAsync(partitionName, groupId, conf));
+                    }
+
+                    // Wait for all consumers to be ready
+                    for (int i = 0; i < partitionMetadata.partitions; i++) {
+                        consumers.putIfAbsent(new TopicPartition(topic, i), futures.get(i).get());
```

Comment:
I think previous comment is not showing up so, adding here:
what if one of the partitioned consumer fails to subscribe? 
eg: out of c1,c2,c3 -> c2 fails so, do we have to close c1 and c3? and should we remove c1 from `consumers` because we already added c1?

Reply:
Oh yes, let me refactor that to make sure we clean all intermediated consumers